tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/type-change-state.html

```diff
@@ -36,26 +36,34 @@
       if (types[i] != types[j]) {
         test(function() {
           var input = document.createElement("input");
+          var expected = "  foo\rbar  ";
           input.type = types[i].type;
           if (types[i].type === "file") {
             assert_throws("INVALID_STATE_ERR", function() {
-              input.value = "  foo\rbar  ";
+              input.value = expected;
             });
             assert_equals(input.value, "");
           } else if (types[j].type === "file") {
-            input.value = "  foo\rbar  ";
+            input.value = expected;
             input.type = types[j].type;  // change state
             assert_equals(input.value, "");
           } else {
-            input.value = "  foo\rbar  ";
+            input.value = expected;
             input.type = types[j].type;  // change state
-            if (types[i].type !== "color" && (types[j].sanitizedValue || types[j].sanitizedValue === "")) {
-              assert_equals(input.value, types[j].sanitizedValue, "input.value should be " + types[j].sanitizedValue + " after change of state");
-            } else if (types[i].sanitizedValue || types[i].sanitizedValue === "") {
-              assert_equals(input.value, types[i].sanitizedValue, "input.value should be " + types[i].sanitizedValue + " after change of state");
-            } else {
-              assert_equals(input.value, "  foo\rbar  ", "input.value should be '  foo\\rbar  ' after change of state");
+
+            // type[i] sanitization
+            if (types[i].sanitizedValue || types[i].sanitizedValue === "") {
```

Comment:
The ordering on this test looks wrong. If I'm reading correctly, `j` represents the destination input type that this input element is changing to, so the value sanitization algorithm should run for the input type described by `types[j]` first.

Reply:
I'm still not sure how to do this right. We have an input sanitization which may set the value to empty string, trims it somehow or replace it with a default value on source input type and then afterwards another that possibly do the same with that value on the destination input type.

Reply:
So the part of the spec that specifies this behaviour is located here https://html.spec.whatwg.org/multipage/input.html#input-type-change, and in the code it's this line:

https://github.com/servo/servo/blob/b7bb6ffbb85aed4e94f36a71fac112ae64d46911/components/script/dom/htmlinputelement.rs#L1016

... which will get called whenever the `value` content attribute is modified.

Reply:
I think we're getting near the expected test behavior, the only thing left is:

```
  ▶ Unexpected subtest result in /html/semantics/forms/the-input-element/type-change-state.html:
  │ FAIL [expected PASS] change state from time to checkbox
  │   → assert_equals: input.value should be '' after change of state expected "" but got "on"
  │ FAIL [expected PASS] change state from time to radio
  │   → assert_equals: input.value should be '' after change of state expected "" but got "on"
```

BTW: There are many tests that pass who should fail (many date / number related input type changes).

Reply:
Ick, this is because the default value for radio and checkbox is "on" instead of the empty string:

https://github.com/servo/servo/blob/b7bb6ffbb85aed4e94f36a71fac112ae64d46911/components/script/dom/htmlinputelement.rs#L396

So, perhaps instead of having a `default` boolean field on the test object, we should have a `defaultValue` string.

Reply:
The more I try to find a nice solution the more I break :/ The additional `defaultValue` is ugly, but at least the tests work now.